### PR TITLE
FTS extensions

### LINKED ISSUE #386: 'l => l.Tags' cannot be converted to SQL. when Join with FreeTextTable
This works fine: (joining an existing query to a FreeTextTable, using inner join)

```cs
return from q in query
   join f in db.FreeTextTable<Link, Guid>(l => l.Tags, authenticatedUserTags) on q.Id equals f.Key
   orderby f.Rank descending
   select q;
```

But this don't: (only changed inner join to left join)

```cs
return from q in query
   from f in db.FreeTextTable<Link, Guid>(l => l.Tags, authenticatedUserTags).Where(f => f.Key == q.Id).DefaultIfEmpty()
   orderby f.Rank descending
   select q;
```

Exception:

```
[LinqException: 'l => l.Tags' cannot be converted to SQL.]
   LinqToDB.Linq.Builder.ExpressionBuilder.ConvertToSql(IBuildContext context, Expression expression, Boolean unwrap) in C:\CodeArt\Submodules\linq2db\Source\Linq\Builder\ExpressionBuilder.SqlBuilder.cs:988
   LinqToDB.Linq.Builder.<>c__DisplayClass28_0.<.ctor>b__0(Expression a) in C:\CodeArt\Submodules\linq2db\Source\Linq\Builder\TableBuilder.cs:190
   System.Linq.WhereSelectEnumerableIterator`2.MoveNext() +223
   System.Linq.Buffer`1..ctor(IEnumerable`1 source) +153
   System.Linq.Enumerable.ToArray(IEnumerable`1 source) +106
   LinqToDB.DataProvider.SqlServer.FreeTextTableExpressionAttribute.SetTable(MappingSchema mappingSchema, SqlTable table, MemberInfo member, IEnumerable`1 expArgs, IEnumerable`1 sqlArgs) in C:\CodeArt\Submodules\linq2db\Source\DataProvider\SqlServer\FreeTextTableExpressionAttribute.cs:28

```

But using string, works fine: (removed lambda)

```cs
return from q in query
   from f in db.FreeTextTable<Link, Guid>("Tags", authenticatedUserTags).Where(f => f.Key == q.Id).DefaultIfEmpty()
   orderby f.Rank descending
   select q;
```

### LINKED ISSUE #1561: [FTS] Add full-text extensions for databases that support it
Right now we ship only FTS extension for SQL Server, but people keep asking for them for other databases:
- sqlite (we already have it in tests) https://github.com/linq2db/linq2db/blob/master/Tests/Linq/Linq/FunctionTests.cs#L471
- mysql https://stackoverflow.com/questions/54408134/how-to-make-full-text-search-with-linq2db

Looks like it makes sense to implement it as a part of linq2db as Sql.Extension is not something easy to implement by user due to lack of documentation/knowledge of linq2db internals

### PR DESCRIPTION
**Pull request by MaceWindu**
fix #1561 

- [x] [SQL Server] FREETEXTTABLE function support
- [x] [SQL Server] CONTAINSTABLE function support
- [x] [SQL Server] FREETEXT predicate support
- [x] [SQL Server] CONTAINS predicate support
- [x] [SQL Server] CONTAINS(PROPERTY()) predicate support
- [x] [SQLite] FTS extensions
- [x] [MySql] FTS extensions
- [x] Move FTS extensions to Sql.Ext.[dbname] extension points
- [ ] ~(maybe) [PostgreSQL] FTS extensions~ postponed till user request

Summary:
- [SQL Server] add new SQL Server extensions for FREETEXTTABLE/CONTAINSTABLE functions and FREETEXT/CONTAINSTEXT predicates with all kinds of parameters support
- [SQL Server] obsolete old `FreeTextTableExpressionAttribute` extension attribute and `Sql.FreeText()` extension
- [SQL Server] set default value of T4 `GenerateSqlServerFreeText` option to `false` to avoid obsoleted `FreeTextTableExpressionAttribute`-related FTS code generation
- fix #386 (by migrating to new extensions, obsoleted `FreeTextTableExpressionAttribute` is still broken and we don't plan to fix it)

### CODE REVIEW

**Comment by MaceWindu** [comment]
@sdanyliv , @it - guys, we still need to decide new extensions naming schema

**Comment by MaceWindu** [comment]
List of new extesions (for naming schema review).

MySQL:
```
// MATCH AGAINST as predicate
Sql.Ext.MySql().Match(search, columns)
Sql.Ext.MySql().Match(modifier, search, columns)
// MATCH AGAINST as relevance
Sql.Ext.MySql().MatchRelevance(search, columns)
Sql.Ext.MySql().MatchRelevance(modifier, search, columns)
```

**Comment by MaceWindu** [comment]
SQLite:
```cs
// match(table) and match(column)
Sql.Ext.SQLite().Match(entityOrColumn, search);
Sql.Ext.SQLite().MatchTable(table, search);

// hidden columns
Sql.Ext.SQLite().RowId(table);
Sql.Ext.SQLite().Rank(table);
```

FTS3 and FTS5 functions:
```cs
// FTS3
Sql.Ext.SQLite().FTS3Offsets(table);
Sql.Ext.SQLite().FTS3MatchInfo(table);
Sql.Ext.SQLite().FTS3MatchInfo(table, format);
Sql.Ext.SQLite().FTS3Snippet(table);
Sql.Ext.SQLite().FTS3Snippet(table, startMatch);
Sql.Ext.SQLite().FTS3Snippet(table, startMatch, endMatch);
Sql.Ext.SQLite().FTS3Snippet(table, startMatch, endMatch, ellipses);
Sql.Ext.SQLite().FTS3Snippet(table, startMatch, endMatch, ellipses, columnIndex);
Sql.Ext.SQLite().FTS3Snippet(table, startMatch, endMatch, ellipses, columnIndex, tokensNumber);

// FTS5
Sql.Ext.SQLite().FTS5bm25(table);
Sql.Ext.SQLite().FTS5bm25(table, params[] weights);
Sql.Ext.SQLite().FTS5Highlight(table, columnIndex, startMatch, endMatch);
Sql.Ext.SQLite().FTS5Snippet(table, columnIndex, startMatch, endMatch, ellipses, tokensNumber);
```

FTS3 and FTS5 commands (DataConnection extensions):
```cs
// FTS3
dc.FTS3Optimize(table);
dc.FTS3Rebuild(table);
dc.FTS3IntegrityCheck(table);
dc.FTS3Merge(table, blocks, segments);
dc.FTS3AutoMerge(table, segments);

// FTS5
dc.FTS5AutoMerge(table, value);
dc.FTS5CrisisMerge(table, value);
dc.FTS5Delete(table, rowid, record);
dc.FTS5DeleteAll(table);
dc.FTS5IntegrityCheck(table);
dc.FTS5Merge(table, value);
dc.FTS5Optimize(table);
dc.FTS5Pgsz(table, value);
dc.FTS5Rank(table, function);
dc.FTS5Rebuild(table);
dc.FTS5UserMerge(table, value);
```

**Comment by MaceWindu** [comment]
SQL Server:

```cs
// FREETEXTTABLE(table, *, ...)
Sql.Ext.SqlServer().FreeTextTable(table, search);
Sql.Ext.SqlServer().FreeTextTable(table, search, top);
Sql.Ext.SqlServer().FreeTextTableWithLanguage(table, search, top, language_name);
Sql.Ext.SqlServer().FreeTextTableWithLanguage(table, search, top, language_name, top);
Sql.Ext.SqlServer().FreeTextTableWithLanguage(table, search, top, language_code);
Sql.Ext.SqlServer().FreeTextTableWithLanguage(table, search, top, language_code, top);

// FREETEXTTABLE(table, (columns), ...)
Sql.Ext.SqlServer().FreeTextTable(table, columns_expr, search);
Sql.Ext.SqlServer().FreeTextTable(table, columns_expr, search, top);
Sql.Ext.SqlServer().FreeTextTableWithLanguage(table, columns_expr, search, top, language_name);
Sql.Ext.SqlServer().FreeTextTableWithLanguage(table,, columns_expr search, top, language_name, top);
Sql.Ext.SqlServer().FreeTextTableWithLanguage(table, columns_expr, search, top, language_code);
Sql.Ext.SqlServer().FreeTextTableWithLanguage(table, columns_expr, search, top, language_code, top);

// CONTAINSTABLE(table, *, ...)
Sql.Ext.SqlServer().ContainsTable(table, search);
Sql.Ext.SqlServer().ContainsTable(table, search, top);
Sql.Ext.SqlServer().ContainsTableWithLanguage(table, search, top, language_name);
Sql.Ext.SqlServer().ContainsTableWithLanguage(table, search, top, language_name, top);
Sql.Ext.SqlServer().ContainsTableWithLanguage(table, search, top, language_code);
Sql.Ext.SqlServer().ContainsTableWithLanguage(table, search, top, language_code, top);

// CONTAINSTABLE(table, (columns), ...)
Sql.Ext.SqlServer().ContainsTable(table, columns_expr, search);
Sql.Ext.SqlServer().ContainsTable(table, columns_expr, search, top);
Sql.Ext.SqlServer().ContainsTableWithLanguage(table, columns_expr, search, top, language_name);
Sql.Ext.SqlServer().ContainsTableWithLanguage(table,, columns_expr search, top, language_name, top);
Sql.Ext.SqlServer().ContainsTableWithLanguage(table, columns_expr, search, top, language_code);
Sql.Ext.SqlServer().ContainsTableWithLanguage(table, columns_expr, search, top, language_code, top);

// FREETEXT(table.*, ..)
Sql.Ext.SqlServer().FreeText(table, search);
Sql.Ext.SqlServer().FreeTextWithLanguage(table, search, language_name);
Sql.Ext.SqlServer().FreeTextWithLanguage(table, search, language_code);

// FREETEXT((columns), ..)
Sql.Ext.SqlServer().FreeText(table, search, param[] columns);
Sql.Ext.SqlServer().FreeTextWithLanguage(table, search, language_name, param[] columns);
Sql.Ext.SqlServer().FreeTextWithLanguage(table, search, language_code, param[] columns);

// CONTAINS(table.*, ..)
Sql.Ext.SqlServer().Contains(table, search);
Sql.Ext.SqlServer().ContainsWithLanguage(table, search, language_name);
Sql.Ext.SqlServer().ContainsWithLanguage(table, search, language_code);

// CONTAINS((columns), ..)
Sql.Ext.SqlServer().Contains(table, search, param[] columns);
Sql.Ext.SqlServer().ContainsWithLanguage(table, search, language_name, param[] columns);
Sql.Ext.SqlServer().ContainsWithLanguage(table, search, language_code, param[] columns);

// CONTAINS(PROPERTY())
Sql.Ext.SqlServer().ContainsProperty(column, property, search);
Sql.Ext.SqlServer().ContainsPropertyWithLanguage(column, property, search, language_name);
Sql.Ext.SqlServer().ContainsPropertyWithLanguage(column, property, search, language_code);
```

**Comment by MaceWindu** [comment]
This is ready for review.
I decided to not add PostgreSQL FTS extensions now as it is another big chunk of work nobody requested yet. Could be added later on user request

File: Source/LinqToDB/DataProvider/MySql/MySqlExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+		/// <summary>
+		/// Applies full-text search condition using MATCH AGAINST predicate against specified full-text columns using default mode (IN NATURAL LANGUAGE MODE).
+		/// Example: MATCH(col1, col2) AGAINST('search query').
+		/// </summary>
+		/// <param name="ext">Extension point.</param>
+		/// <param name="search">Full-text search condition.</param>
+		/// <param name="columns">Full-text columns that should be queried.</param>
+		/// <returns>Returns <c>true</c> if full-text search found matching records.</returns>
+		[Sql.Extension("MATCH({columns, ', '}) AGAINST ({search})", IsPredicate = true, ServerSideOnly = true)]
+		public static bool Match(this IMySqlExtensions ext, [ExprParameter("search")] string search, [ExprParameter("columns")] params object[] columns)
```
Comment (sdanyliv) [inline Source/LinqToDB/DataProvider/MySql/MySqlExtensions.cs]: There is no one place. But name of parameter is not required - original parameter name will be used

File: Tests/T4.Linq/SqlServer.generated.cs [hunk 2]
```diff
@@ -1298,11 +1256,11 @@ public static Territory Region(this Region obj, IDataContext db)
...
 	/// <summary>
-	/// Database       : TestData
-	/// Data Source    : DBHost2\SQLSERVER2014
-	/// Server Version : 12.00.6024
+	/// Database       : TestData2014
+	/// Data Source    : .
+	/// Server Version : 13.00.5233
 	/// </summary>
-	public partial class TestDataDB : LinqToDB.Data.DataConnection
+	public partial class TestData2014DB : LinqToDB.Data.DataConnection
```
Comment (sdanyliv) [inline Tests/T4.Linq/SqlServer.generated.cs]: Looks like we have to set name of DataConnection class explicitly.

**Comment by MaceWindu** [comment]
Whoops, forgot to address review comments, will do in master

**Pull request merged**